core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -10383,6 +10423,310 @@ NABoolean Insert::isUpsertThatNeedsMerge(NABoolean isAlignedRowFormat, NABoolean
+
+  BindScope *upsertScope = bindWA->getCurrentScope();
+  // Create a new BindScope, to encompass the new nodes 
+  // upsert(left_join(input_scan, tuple))
+  // and any inlining nodes that will be created. Any values the upsert
+  // and children will need from src will be marked as outer references in that
+  // new BindScope. We assume that "src" is already bound.
+  ValueIdSet currOuterRefs = bindWA->getCurrentScope()->getOuterRefs();
+  // Save the current RETDesc.
+  RETDesc *prevRETDesc = bindWA->getCurrentScope()->getRETDesc();
+
+  // columns of the target table
+  const ValueIdList &tableCols = updateToSelectMap().getTopValues();
+  const ValueIdList &sourceVals = updateToSelectMap().getBottomValues();
+
+  // create a Join node - left join of the base table columns with the columns to be upserted.
+  // columns of the target table
+  CMPASSERT(child(0)->nodeIsBound());
+  
+  Scan * targetTableScan =
+    new (bindWA->wHeap())
+    Scan(CorrName(getTableDesc()->getCorrNameObj(), bindWA->wHeap()));
+
+ 
+  //join predicate between source columns and target table.
+  ItemExpr * keyPred = NULL;
+  ItemExpr * keyPredPrev = NULL;
+  BaseColumn* baseCol;
+  ColReference * targetColRef;
+  int predCount = 0;
+  ValueIdSet newOuterRefs;
+  ItemExpr * pkeyValPrev;
+  ItemExpr * pkeyVals = NULL;
+  for (CollIndex i = 0; i < tableCols.entries(); i++)
+    {
+      baseCol = (BaseColumn *)(tableCols[i].getItemExpr()) ;
+      if (baseCol->getNAColumn()->isSystemColumn())
+	continue;
+
+      targetColRef = new(bindWA->wHeap()) ColReference(
+						       new(bindWA->wHeap()) ColRefName(
+										       baseCol->getNAColumn()->getFullColRefName(), bindWA->wHeap()));
+    
+
+      if (baseCol->getNAColumn()->isClusteringKey())
+	{
+	  // create a join/key predicate between source and target table,
+	  // on the clustering key columns of the target table, making
+	  // ColReference nodes for the target table, so that we can bind
+	  // those to the new scan
+	  keyPredPrev = keyPred;
+	  keyPred = new (bindWA->wHeap())
+	    BiRelat(ITM_EQUAL, targetColRef, 
+		    sourceVals[i].getItemExpr(),
+		    baseCol->getType().supportsSQLnull());
+	  predCount++;
+	  if (predCount > 1) 
+	    {
+	      keyPred = new(bindWA->wHeap()) BiLogic(ITM_AND,
+						     keyPredPrev,
+						     keyPred);  
+	    }
+	  pkeyValPrev = pkeyVals;
+    
+	  pkeyVals = tableCols[i].getItemExpr();
+	  
+	  if (i > 0) 
+	    {
+	      pkeyVals = new(bindWA->wHeap()) ItemList(pkeyVals,pkeyValPrev);
+      
+	    }
+	}
+     
+    }
+ 
+  // Map the table's primary key values to the source lists key values
+  ValueIdList tablePKeyVals = NULL;
+  ValueIdList sourcePKeyVals = NULL;
+  
+  pkeyVals->convertToValueIdList(tablePKeyVals,bindWA,ITM_ITEM_LIST);
+  updateToSelectMap().mapValueIdListDown(tablePKeyVals,sourcePKeyVals);
+  
+
+
+  Join *lj = new(bindWA->wHeap()) Join(child(0),targetTableScan,REL_LEFT_JOIN,keyPred);
+  lj->doNotTransformToTSJ();	  
```

Comment:
I assume this is because we want to use the old state of the target table, before we applied any updates. But, what if I do an upsert of 1000 rows into a table with 10 billion rows? Do we really want to read 10 billion rows with all their columns into a hash join or merge join? The merge would be vastly more efficient for this case - or am I missing something?

Another way to see the old state of the table would be to add a sort node above the join, so in this case we would only need to sort 1000 rows and we may need to do a sort anyway for the sequence. So, maybe just add code that forces a sort operator for the required order of the sequence node (at least forces it if we use a nested join or a merge join w/o a sort).